compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.td

```diff
@@ -0,0 +1,173 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_CODEGEN_DIALECT_IREEGPUOPS
+#define IREE_CODEGEN_DIALECT_IREEGPUOPS
+
+include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUDialect.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+include "mlir/IR/OpAsmInterface.td"
+include "mlir/IR/OpBase.td"
+
+def MemrefTypeAttr : TypeAttrBase<"MemRefType", "Memref type attribute">;
+
+//===----------------------------------------------------------------------===//
+// ShuffleTensorOp
+//===----------------------------------------------------------------------===//
+
+def IREEGPU_ShuffleTensorOp : Op<IREEGPU_Dialect, "shuffle_tensor", [
+    Pure,
+    AttrSizedOperandSegments
+    ]> {
+  let summary = "Shuffles a private tensor across a shared allocation";
+  let description = [{
+    This op is designed to represent a shuffle of private tensor data
+    collectively held across a set of workers. This operation naturally arises
+    when combining the regions of producer-consumer `scf.forall` operations
+    that share a mapping type and worker count.
+
+    For example, consider the following pair of parallel loops.
+    ```mlir
+      %0 = scf.forall (%idy, %idx) in (2, 32) shared_outs(%init = %empty) -> (tensor<4x128xf32>) {
+        %in = ...
+        %2 = affine.apply #affine_map<(d0) -> (d0 * 2)> (%idy)
+        %3 = affine.apply #affine_map<(d0) -> (d0 * 4)> (%idx)
+        scf.forall.in_parallel {
+          tensor.parallel_insert_slice %in into %init[%2, %3] [2, 4] [1, 1]
+            : tensor<2x4xf32> into tensor<4x128xf32>
+        }
+      } {mapping = [#gpu.thread<y>, #gpu.thread<x>]}
+      %1 = scf.forall (%idy, %idx) in (8, 8) -> (tensor<128x128xf32>) {
+        %4 = affine.apply #affine_map<(d0) -> (d0 * 16)> (%idx)
+        %extracted_slice = tensor.extract_slice %0[0, %4] [4, 16] [1, 1]
+          : tensor<4x128xf32> to tensor<4x16xf32>
+        ...
+      } {mapping = [#gpu.thread<y>, #gpu.thread<x>]}
+    ```
+
+    Because these loops share the same worker type and total count, the bodies
+    of these two loops can be merged with a barrier and a shuffle where the
+    boundary of the loops currently is.
+
+    ```mlir
+      %0 = scf.forall (%idy, %idx) in (8, 8) -> (tensor<4x128xf32>) {
+        %ids = affine.delinearize_index %idy * 8 + %idx to (2, 32) : index
+        %in = ...
+        %2 = affine.apply #affine_map<(d0) -> (d0 * 2)> (%ids#0)
+        %3 = affine.apply #affine_map<(d0) -> (d0 * 4)> (%ids#1)
+        %4 = affine.apply #affine_map<(d0) -> (d0 * 16)> (%idx)
+        %slice = iree_gpu.shuffle_tensor %in[%2, %3] [2, 4] [1, 1] to %empty[0, %4] [4, 16] [1, 1]
```

Comment:
So trying to understand the semantics. When you say write of the source slice, you are essentially refering to 
```
tensor.insert_slice
gpu.barrier
tensor.extract_slice.
```

So what is the "source" here?  Also it might just be better to use a "memref" type as the intermediate. Eventually you want to lower to a write to a memref and read from the memref... If you dont use a memref, you need this tensor to be defined outside the loop and be passed in maybe through iter_args, which seems more involved...

Reply:
I initially tried making it a memref but couldn't figure out how to lower it to `memref.alloc` without using `vector.transfer_read/write` ops. Maybe I should disallow dynamically sized shuffles actually and that would fix the problem. I think I'll do that and change the fusion function to require static. In the future we can figure out how dynamic would work.

Reply:
After some more thought, I don't mind making this change, but I think there is good justification for keeping this as a tensor + destination at this level. The semantics of [tensor.parallel_insert_slice](https://mlir.llvm.org/docs/Dialects/TensorOps/#tensorparallel_insert_slice-tensorparallelinsertsliceop) are:
```
If an index is updated exactly once, the value contained at that index in the resulting
tensor will be equal to the value at a corresponding index of a slice that was used for
the updated. If an index is not updated at all, its value will be equal to the one in the
original tensor.
```

So to replace the "intermediate tensor" (the destination of the forall) with a memref TypeAttr means we need to match the init for the forall to a `tensor.empty` and/or guarantee that all of the threads combine to overwrite the entire destination. It's probably not a case we'll see soon, but I could see it appearing with specific fusions like `tensor.concat`.

Reply:
The other two options I see are
1. Immediately convert the destination of the forall with `bufferization.to_memref`.
2. Allow this shuffle op to take both tensors and memrefs and do (1) after the fusion of forall ops.

Reply:
I realized my concat example doesn't work unless the intermediate allocation is a memref so I changed it to a memref and added a comment.